### DISCUSSION TITLE
Using existing Package [timacdonald/json-api]

### DISCUSSION THREAD

**Discussion by lloricode** [dedoc/scramble#185] [Ideas]
Hi, we are using [timacdonald/json-api](https://github.com/timacdonald/json-api), it uses laravel API resource but it provide `toAttribute()` instead of `toArray()`, and notice did not generate response base on API resource, here my demo site https://ecommerce.lloricode.com/docs/api and i just install the library without API annotation.

Can you add support this on `timacdonald/json-api`? or is there a way to override its auto-generation?

Best regards,

**Comment by romalytvynenko** [comment]
@lloricode you can explore Scramble's Extensions API to add a support for this package.

https://scramble.dedoc.co/developers/extensions

**Reply by lloricode** [reply]
thanks, will check this later

**Reply by lloricode** [reply]
hi @romalytvynenko I was a struggle to achieve the exact format, my goal is to have same as [json:api](https://jsonapi.org/) specs, but I can only make like this

![Screenshot_2023-08-20_14-07-12](https://github.com/dedoc/scramble/assets/8251344/378ba138-4790-48c4-8df3-6a9831fc3bf0)

with extension:

```php

<?php

declare(strict_types=1);

namespace App;

use Dedoc\Scramble\Extensions\TypeToSchemaExtension;
use Dedoc\Scramble\Support\Generator\Reference;
use Dedoc\Scramble\Support\Type\ObjectType;
use Dedoc\Scramble\Support\Type\Type;
use TiMacDonald\JsonApi\JsonApiResource;

class CustomTypeToSchemaExtension extends TypeToSchemaExtension
{
    public function shouldHandle(Type $type): bool
    {
        return $type instanceof ObjectType
            && $type->isInstanceOf(JsonApiResource::class);
    }

    public function toSchema(Type $type): mixed
    {
        $this->infer->analyzeClass($type->name);

        $array = $type->getMethodDefinition('toAttributes')
            ->type
            ->getReturnType();

        return $this->openApiTransformer->transform($array);
    }

    public function reference(ObjectType $type): Reference
    {
        return new Reference('schemas', $type->name, $this->components);
    }
}
```

expecting something like this

```json
{
  "data": [
    {
      "id": "samsung-galaxy-s21",
      "type": "products",
      "attributes": {
        "parent_sku": "f9aa4c5e-bcac-3335-badb-12dfd7e998fe",
        "name": "Samsung Galaxy S21",
        "description": "........",
        "status": "enabled"
      },
      "relationships": {},
      "meta": {},
      "links": {}
    }
  ]
}
```

I just like atleast to wrap it with:

```json
{
  "data": [
    {
      "id": "string",
      "type": "TYPE of RESOURCE",
      "attributes": {
      },
      "relationships": {},
      "meta": {},
      "links": {}
    }
  ]
}
```

so the API consumer will not confuse of the API response

**Reply by romalytvynenko** [reply]
@lloricode This is a great starting point!

To make what you need, most likely you'd want to "construct" the resulting type yourself. 

As a reference, you can check how Scramble implements pagination responses/schemas: https://github.com/dedoc/scramble/blob/aa05297338260de10cf03c314897b89371c9643d/src/Support/TypeToSchemaExtensions/LengthAwarePaginatorTypeToSchema.php#L20

Notice how I construct the resulting type there.

**Reply by lloricode** [reply]
looks good to me, will try this

thank you

**Reply by lloricode** [reply]
Hi @romalytvynenko, I can not achieve and it seems to difficult to make it,

**Reply by romalytvynenko** [reply]
@lloricode should be doable - the only thing left for you is to construct a type in an extension.

Feel free to share code or details

**Reply by lloricode** [reply]
nothing much changes, the `toResponse()` seems not working without `toSchema()`
